modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+    var buff: bufferType = nil;
+    var owned: bool = true;
+
+    proc string(s: string, owned: bool = true) {
+      this.owned = owned;
+      const sRemote = s.locale.id != chpl_nodeID;
+      if !owned && sRemote {
+        // TODO: Should I just ignore the supplied value of owned instead of
+        //       halting?
+        halt("Must take ownership of a copy of a remote string");
+      }
+      var localBuff: bufferType = nil;
+      const sLen = s.len;
+      // Dont need to do anything if s is an empty string
+      if sLen != 0 {
+        localBuff = if sRemote
+          then copyRemoteBuffer(s.locale.id, s.buff, sLen)
+          else s.buff;
+      }
+      // We only need to copy when the buffer is local
+      // TODO: reinitString does more checks then we actually need but is still
+      // correct. Is it worth just doing the proper copy here?
+      this.reinitString(localBuff, sLen, sLen+1, needToCopy=sRemote);
+    }
 
-  inline proc +(x: numeric, s: string)
-    return x:string + s;
+    // This constructor can cause a leak if owned = false and needToCopy = true
+    proc string(buff: bufferType, length: int, size: int,
+                owned: bool = true, needToCopy: bool = true) {
+      this.owned = owned;
+      this.reinitString(buff, length, size, needToCopy);
+    }
 
-  inline proc +(s: string, x: enumerated)
-    return s + x:string;
+    proc ref ~string() {
+      if owned && !this.isEmptyString() {
+        on this {
+          chpl_mem_free(this.buff);
+        }
+      }
+    }
 
-  inline proc +(x: enumerated, s: string)
-    return x:string + s;
+    // This is assumed to be called from this.locale
+    proc ref reinitString(buf: bufferType, s_len: int, size: int,
+                          needToCopy:bool = true) {
+      if debugStrings then
+        chpl_debug_string_print("in string.reinitString()");
+
+      if this.isEmptyString() {
+        if (s_len == 0) || (buf == nil) then return; // nothing to do
+      }
+
+      // If the this.buff is longer than buf, then reuse the buffer if we are
+      // allowed to (this.owned == true)
+      if s_len != 0 {
+        if needToCopy {
+          if !this.owned || s_len+1 > this._size {
+            // If the new string is too big for our current buffer or we dont
+            // own our current buffer then we need a new one.
+            if this.owned && !this.isEmptyString() then
+              chpl_mem_free(this.buff);
+            // TODO: should I just allocate 'size' bytes?
+            const allocSize = chpl_mem_goodAllocSize((s_len+1).safeCast(size_t));
+            this.buff = chpl_mem_alloc(allocSize,
+                                       CHPL_RT_MD_STR_COPY_DATA):bufferType;
+            this.buff[s_len] = 0;
+            this._size = allocSize.safeCast(int);
+            // We just allocaed a buffer, make sure to free it later
+            this.owned = true;
+          }
+          memmove(this.buff, buf, s_len.safeCast(size_t));
```

Comment:
Why is this memmove (vs memcpy)? Later (when I get further down) are you going to call reinitString with buf being an alias for part of this.buff? Would it be better to have a separate function that did that?